File: Programmers/거리두기 확인하기.py
# 대기실은 5개이며, 각 대기실은 5x5 크기입니다.
# 거리두기를 위하여 응시자들 끼리는 맨해튼 거리1가 2 이하로 앉지 말아 주세요.
# 단 응시자가 앉아있는 자리 사이가 파티션으로 막혀 있을 경우에는 허용합니다
def solution(places):
    answer = []
    for place in places:
        answer.append(find_illegal(place))
    return answer


def find_illegal(place):
    direction = [[1, 0], [-1, 0], [0, 1], [0, -1]]

    find = []

    def DFS(x, y, start_x, start_y):
        visited[x][y] = True
        answer = True
        for dir in direction:
            dx = x + dir[0]
            dy = y + dir[1]
            if 0 <= dx < 5 and 0 <= dy < 5:
                if visited[dx][dy] == False and calculate_distance(start_x, start_y, dx, dy) <= 2:
                    if place[dx][dy] == "P":
                        return False
                    else:
                        answer = DFS(dx, dy, start_x, start_y)
            if answer == False:
                return False
        return answer

    for i in range(5):
        for k in range(5):
            if place[i][k] == "P":
                visited = []
                for row in range(5):
                    temp = []
                    for col in range(5):
                        if place[row][col] == "X":
                            temp.append(True)
                        else:
                            temp.append(False)
                    visited.append(temp)
                temp = DFS(i, k, i, k)
                if temp == False:
                    return 0
    return 1


def calculate_distance(r1, c1, r2, c2):
    return abs(r1 - r2) + abs(c1 - c2)

File: Programmers JS/거리두기 확인하기.js
var answer = [];
var direction = [
  [-1, 0],
  [1, 0],
  [0, 1],
  [0, -1],
];
function solution(places) {
  for (let item of places) {
    let flag = 0;
    for (let i = 0; i < 5; i++) {
      for (let j = 0; j < 5; j++) {
        if (item[i][j] === "P") {
          let visited = [...item.map((el) => el.split(""))];
          let result = DFS(visited, [i, j], [i, j], 0);
          // console.log("item",item);
          // console.log(i,j);
          // console.log(result)
          if (!result) {
            flag = 1;
          }
        }
      }
    }
    if (flag === 1) {
      answer.push(0);
    } else {
      answer.push(1);
    }
  }

  return answer;
}

function DFS(visited, start, current, depth) {
  let result = true;
  visited[current[0]][current[1]] = "X";

  for (let dir of direction) {
    let dx = current[0] + dir[0];
    let dy = current[1] + dir[1];

    if (0 <= dx && dx < 5 && 0 <= dy && dy < 5) {
      // console.log("why",[dx,dy],depth)
      // console.log("visited",visited[dx][dy],depth)
      if (visited[dx][dy] !== "X" && get_distance(start, [dx, dy]) <= 2) {
        // console.log("check",start,current,[dx,dy])
        if (visited[dx][dy] === "P") {
          // console.log("dxdy",start,[dx,dy])
          return false;
        }
        result = DFS(visited, start, [dx, dy], depth + 1);
      }
    }
  }
  return result;
}

function get_distance(t1, t2) {
  return Math.abs(t1[0] - t2[0]) + Math.abs(t1[1] - t2[1]);
}
